lib/pos_staking_rewards.go

```diff
@@ -1,5 +1,271 @@
+	// their staking rewards over the long run. Validators can opt out of it by disabling reward restaking on their own StakeEntry.
+	// 2. It simplifies the validator commission distribution code by re-using the same code path for distributing staking
+	// rewards. By requiring the validator to already have a StakeEntry for themselves if they want to restake their commissions,
+	// this approach allows us to avoid manually creating new StakeEntries for the validator specifically for restaking commissions.
+	//
+	// TODO: The downside of the above is that it couples the restaking behavior for validator commissions and the validator's own
+	// staking reward. This is fine though, because if the validator wants to restake their own rewards but not their commissions, then
+	// they can stake to themselves using a separate wallet and only enable reward restaking for that StakeEntry.
+	//
+	// If the above isn't desired the behavior, then we can alternatively always pay out validator's commission directly to their wallet.
+	return bav.distributeStakingReward(validatorPKID, validatorPKID, commissionAmount)
+}
+
+const (
+	_basisPoints     = uint64(10000)                    // 1e4
+	_nanoSecsPerYear = uint64(365) * 24 * 60 * 60 * 1e9 // 365 days * 24 hours * 60 minutes * 60 seconds * 1e9 nanoseconds
+)
+
+var (
+	_basisPointsAsInt       = big.NewInt(int64(_basisPoints))
+	_basisPointsAsFloat     = NewFloat().SetUint64(_basisPoints)
+	_nanoSecsPerYearAsFloat = NewFloat().SetUint64(_nanoSecsPerYear)
+)
+
+func convertBigFloatToBigInt(float *big.Float) *big.Int {
+	floatAsInt, _ := float.Int(nil)
+	return floatAsInt
+}
+
+func convertAPYBasisPointsToFloat(apyBasisPoints uint64) *big.Float {
+	apyBasisPointsAsFloat := NewFloat().SetUint64(apyBasisPoints)
+	return apyBasisPointsAsFloat.Quo(apyBasisPointsAsFloat, _basisPointsAsFloat)
+}
+
+func computeFractionOfYearAsFloat(nanoSecs uint64) *big.Float {
+	nanoSecsAsFloat := NewFloat().SetUint64(nanoSecs)
+	return nanoSecsAsFloat.Quo(nanoSecsAsFloat, _nanoSecsPerYearAsFloat)
+}
+
+// computeStakingReward uses float math to compute the compound interest on the stake amounts based on the
+// elapsed time since the last staking reward distribution and the APY.
+//
+// It produces the result for: stakeAmount * [e ^ (apy * elapsedTime / 1 year) - 1]
+func computeStakingReward(stakeAmount *uint256.Int, elapsedFractionOfYear *big.Float, apy *big.Float) *big.Float {
+	stakeAmountFloat := NewFloat().SetInt(stakeAmount.ToBig())
+	growthExponent := NewFloat().Mul(elapsedFractionOfYear, apy)           // apy * elapsedTime / 1 year
+	growthMultiplier := BigFloatPow(bigE, growthExponent)                  // e ^ (apy * elapsedTime / 1 year)
+	finalStakeAmount := NewFloat().Mul(stakeAmountFloat, growthMultiplier) // stakeAmount * [e ^ (apy * elapsedTime / 1 year) - 1]
+	return finalStakeAmount.Sub(finalStakeAmount, stakeAmountFloat)        // stakeAmount * [e ^ (apy * elapsedTime / 1 year) - 1]
```

Comment:
Technically should the comment on line 259 be `stakeAmount * [e ^ (apy *elapsedTime / 1 year)` since we don't sub the original amount until line 260?

Reply:
You're right. Good catch.